avoid capturing the value of the as-response protocol fn

scgilardi opened the pull request:
Before this change, the `:as-response` entry in `default-functions` captures
the _value_ of the protocol fn `as-response` when `core.clj` is loaded. If
the `Represenation` protocol is later extended, those extensions are _not_
seen by the captured fn. After this change, extensions made after
`core.clj` is loaded _are_ available to be used in processing the
resource.

The usual way to avoid this kind of capture when storing a reference to
a Clojure function is to store the function's var itself rather than that
var's value. Doing that in this case causes tests to fail with
exceptions related to how the function/var is called. In contrast, the
proposed solution of using an anonymous wrapper function works well.

ordnungswidrig commented:
Interesting observation. I think your solution is spot on. Can you please add a test case that fails with the current implementation and works with the proposed changes?

scgilardi commented:
Thanks! I added a test.

ordnungswidrig commented on src/liberator/util.clj:
```diff
@@ -5,7 +5,7 @@
 
 (defn make-function [x]
-  (if (or (fn? x) (keyword? x)) x (constantly x)))
+  (if (or (fn? x) (keyword? x) (var? x)) x (constantly x)))
```
I think there's a bug for the case the you define pass a constant value as a var like in

```
(define body "This is the body")

(defresource test
  :handle-ok #'body)
```

This use case might be a bit stretched but I we should get it right.

ordnungswidrig commented on test/test_representation.clj:
```diff
@@ -77,3 +79,18 @@
+                            :handle-ok (->TestRecord))))
+             (catch Exception e
+               {:status 500}))
+        => (contains {:status 200})))
```
I think the test does not need to catch the exception.

scgilardi commented:
updated based on your feedback. thanks.

scgilardi commented:
Essentially the same underlying fix was merged with #232 . All that remains of this PR is the test. I've updated the checkin comment and rebased on master. If you'd like to include the test, please merge. Otherwise, please close. Thanks.

ordnungswidrig commented:
Thanks for the patch!
